### PR TITLE
[Feature]  Add Operator Product 

### LINKED ISSUE #153: [Feature]  Add  operator_product
In order to solve #145, we must compute:

$$
S(\rho) = \sum_i K_i \rho K^{\dagger}_i 
$$

To compute each product  $K_i \rho K^{\dagger}_i$, operators must be allowed to multiply each other.
With this in mind, we will implement the `apply_ope_ope()` function.

### LINKED ISSUE #154: [Feature] Allow different size operators to multiply each other
In order to solve #153 in a more general way, we must allow different size operators to multiply each other. 
With this in mind, we will implement the `promote_ope()`  function.

Promoting an operator means scaling up his dimension to the whole system dimension. 
For instance, if we want to apply a `X` gate targeting the first qubit of a two qubits state in the density matrix formalism, we must compute:

$$
X^{prom} \rho = (X_1 \otimes I_{2}) \rho_{12}
$$

So we need to compute the promote operator $X^{prom} = X_1 \otimes I_{2}$ of `X` by using `promote_ope()` .

### PR DESCRIPTION
**Pull request by EthanObadia**
Add `promote_operator()` function to pyqtorch/utils.py
Add `operator_product()` function to pyqtorch/apply.py

As @jpmoutinho said [here](https://github.com/pasqal-io/pyqtorch/pull/156#discussion_r1557102350), this version of `promote_operator()` can be optimize. So, I believe that its optimization will be the object of an other PR later.

Closes #153
Closes #154

### CODE REVIEW

File: pyqtorch/apply.py
Comment (Roland-djee) [inline pyqtorch/apply.py:95]: Would it make sense to add this function to `utils`? I see it reused many times throughout PyQ. It seems like a bit of a waste to redefine it every time.

File: pyqtorch/apply.py
Comment (Roland-djee) [inline pyqtorch/apply.py:109]: Same.

File: pyqtorch/utils.py
Comment (Roland-djee) [inline pyqtorch/utils.py:166]: You can maybe pass the whole `qubit_support` as opposed to deducing it from `n_qubits` ?

File: pyqtorch/utils.py
Comment (Roland-djee) [inline pyqtorch/utils.py:167]:
```suggestion
    for qubit in qubits_support:
```

**Review by Roland-djee — COMMENTED** [review]
Looking good. Few comments from me.

File: pyqtorch/utils.py
Comment (jpmoutinho) [inline pyqtorch/utils.py:148]: Are you sure this is the best way to apply a given operator? Padding the smaller one with identities makes it quite expensive. I feel like there should be a smart way to apply the small operator on the correct subspace. But I am not sure, I have never really done any density matrix coding.

Comment (jpmoutinho) [reply pyqtorch/utils.py:148]: Maybe this reference might have some cool ideas: https://arxiv.org/abs/2103.13962
And they have some code here: https://github.com/cmendl/density-matrix-bloch-qcircuit/tree/master

But tbh if there is not some easily implementable idea in that paper or somewhere else, I think your approach is also ok for an initial prototype. We can then optimize it later.

Comment (EthanObadia) [reply pyqtorch/utils.py:148]: In theoretical terms, that's how it's done, but indeed I hadn't considered the complexity. So, I'll think about a less expensive version.
However, before making any major modifications to a function, I'd like to finish the implementation of the Kraus operator (#148) completely, even if it means revisiting each of the functions later.
Also, if you want to discuss the` promote_ope()` function, please do so on this PR (#155). Here, it's just there for `apply_ope_ope()` to work and it's not the final version.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:95]: For now, I'm using this function exclusively here, but indeed, it could be a good idea to provide access to this function more generally for future use or for user manipulation. I'll move it to `pyqtorch/utils.py` .

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:109]: For now, I'm using this function here and indirectly in the `density_mat()` function, so I'll move it to `pyqtorch/utils.py`. 
Once the merge is complete, I'll create a PR to use it in `density_mat()`. Or is it possible that I can make modifications here even if the PR is not on this subject ?

File: pyqtorch/utils.py
Comment (EthanObadia) [reply pyqtorch/utils.py:166]: `qubit_support` is defined in `Primitive` as : `qubit_support = (target,)` and not as the list of all the qubit using in the system.
So maybe changing the name variable as: `qubits` is more appropriate right ?

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py]: Done.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py]: Done.

File: pyqtorch/utils.py
Comment (EthanObadia) [reply pyqtorch/utils.py:166]: Done, check #155.

File: pyqtorch/utils.py
Comment (EthanObadia) [reply pyqtorch/utils.py:167]: Done, check #155

File: pyqtorch/utils.py
Comment (EthanObadia) [inline pyqtorch/utils.py:148]: This PR is only about `apply_ope_ope()`. The function `promote_ope()` is just here to allow the test. If you want to review `promote_ope()`, please use #155.

File: pyqtorch/apply.py
Comment (Roland-djee) [inline pyqtorch/apply.py:64]: Maybe `apply_op_to_op` ? Merely a suggestion.

File: pyqtorch/apply.py
Comment (Roland-djee) [inline pyqtorch/apply.py:83]: I think you can combine these conditions because if `n_qubits_1 > n_qubits_2` -> `n_qubits_1 != n_qubits_2`.

File: pyqtorch/apply.py
Comment (Roland-djee) [inline pyqtorch/apply.py:88]: Same

File: pyqtorch/primitive.py
Comment (Roland-djee) [inline pyqtorch/primitive.py:44]:
```suggestion
    def unitary(self, values: dict[str, Tensor] | Tensor = dict()) -> Tensor:
```

File: pyqtorch/primitive.py
Comment (Roland-djee) [inline pyqtorch/primitive.py:47]:
```suggestion
    def forward(self, state: Tensor, values: dict[str, Tensor] | Tensor = dict()) -> Tensor:
```

File: pyqtorch/primitive.py
Comment (Roland-djee) [inline pyqtorch/primitive.py:52]:
```suggestion
    def dagger(self, values: dict[str, Tensor] | Tensor = dict()) -> Tensor:
```

File: pyqtorch/utils.py
Comment (Roland-djee) [inline pyqtorch/utils.py:184]: It should be indented ?

File: tests/test_digital.py
Comment (Roland-djee) [inline tests/test_digital.py:335]: I would create a module-wide global variable since it is reused in two tests:

```python
GATESET = [I, X, Y, Z, H, T, S]
```

**Review by Roland-djee — COMMENTED** [review]
Nice work @EthanObadia looking good. Will approve after you've addressed these very last comments.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:83]: Done.

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py:335]: Changed.

File: pyqtorch/primitive.py
Comment (EthanObadia) [reply pyqtorch/primitive.py:52]: Done.

File: pyqtorch/utils.py
Comment (EthanObadia) [reply pyqtorch/utils.py:184]: Changed.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:64]: It makes sense to change the name. 
Perhaps `opmul`, akin to `torch.matmul()`, would be better as it is shorter and more explicit. 
However, if we aim to maintain the same syntax as the `apply_operator()` function, then `apply_op_to_op` would be more suitable. 
Which name should we stick with?

Comment (Roland-djee) [reply pyqtorch/apply.py:64]: Hm, I would be inclined to use `op_mul` but we can decide together with @dominikandreasseitz

**Review by dominikandreasseitz — APPROVED** [review]

File: pyqtorch/utils.py
Comment (EthanObadia) [reply pyqtorch/utils.py:147]: Changed.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:64]: Done.

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py:329]: Done.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:86]: Indeed, there is an issue if we want to switch from a batch size of 2 to 3 because the function will repeat a tensor with `batch_size = 2` three times, thus creating one with `batch_size = 6`. 
However, in our use case of this function, such a scenario is impossible since it involves computing a product between a gate always with `batch_size = 1` and a batch_size-invariant density matrix.

 So, I'm not sure if a change is necessary. What do you think?

Comment (dominikandreasseitz) [reply pyqtorch/apply.py:86]: yeah thats true. i think we can keep it then

**Review by dominikandreasseitz — APPROVED** [review]
lgtm

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:86]: Finally i fixed it by doing this version:
```py
def operator_product(operator_1: Tensor, operator_2: Tensor, target: int) -> Tensor:
    """
    Compute the product of two operators.

    Args:
        operator_1 (Tensor): The first operator.
        operator_2 (Tensor): The second operator.
        target (int): The target qubit index.

    Returns:
        Tensor: The product of the two operators.
    """

    n_qubits_1 = int(log2(operator_1.size(1)))
    n_qubits_2 = int(log2(operator_2.size(1)))
    batch_size_1 = operator_1.size(-1)
    batch_size_2 = operator_2.size(-1)
    if n_qubits_1 > n_qubits_2:
        operator_2 = promote_operator(operator_2, target, n_qubits_1)
    elif n_qubits_1 < n_qubits_2:
        operator_1 = promote_operator(operator_1, target, n_qubits_2)
    if batch_size_1 > batch_size_2:
        operator_2 = operator_2.repeat(1, 1, batch_size_1)[:,:,:batch_size_1]
    elif batch_size_2 > batch_size_1:
        operator_1 = operator_1.repeat(1, 1, batch_size_2)[:,:,:batch_size_2]


    return batch_last(torch.bmm(batch_first(operator_1), batch_first(operator_2)))
```
But maybe this version is more efficient in term of complexity as this case is rare:
```python3
def operator_product(operator_1: Tensor, operator_2: Tensor, target: int) -> Tensor:
    """
    Compute the product of two operators.

    Args:
        operator_1 (Tensor): The first operator.
        operator_2 (Tensor): The second operator.
        target (int): The target qubit index.

    Returns:
        Tensor: The product of the two operators.
    """

    n_qubits_1 = int(log2(operator_1.size(1)))
    n_qubits_2 = int(log2(operator_2.size(1)))
    batch_size_1 = operator_1.size(-1)
    batch_size_2 = operator_2.size(-1)
    if n_qubits_1 > n_qubits_2:
        operator_2 = promote_operator(operator_2, target, n_qubits_1)
    elif n_qubits_1 < n_qubits_2:
        operator_1 = promote_operator(operator_1, target, n_qubits_2)
    if batch_size_1 > batch_size_2:
        if batch_size_2 == 1:
            operator_2 = operator_2.repeat(1, 1, batch_size_1)
        else:
            operator_2 = operator_2.repeat(1, 1, batch_size_1)[:,:,:batch_size_1]
    elif batch_size_2 > batch_size_1:
        if batch_size_1 == 1:
            operator_1 = operator_1.repeat(1, 1, batch_size_2)
        else:
            operator_1 = operator_1.repeat(1, 1, batch_size_2)[:,:,:batch_size_2]

    return batch_last(torch.bmm(batch_first(operator_1), batch_first(operator_2)))

```
What do you think ?

Comment (Roland-djee) [reply pyqtorch/apply.py:86]: I guess the first one is already quite clear. If it takes into account the second case then it is fine. I would rename `operator_1` -> `op1` since it is quite obvious what we are dealing with.

Comment (EthanObadia) [reply pyqtorch/apply.py:86]: Done.

File: tests/test_digital.py
Comment (Roland-djee) [inline tests/test_digital.py:331]:
```suggestion
def test_promote(operator: list) -> None:
```

File: tests/test_digital.py
Comment (Roland-djee) [inline tests/test_digital.py:343]:
```suggestion
def test_operator_product(operator: list) -> None:
```

**Review by Roland-djee — APPROVED** [review]
Hey @EthanObadia good job ! Looking good to me. Minor typing comments.

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py:331]: Done.

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py:343]: Changed.

File: pyqtorch/apply.py
Comment (gvelikova) [inline pyqtorch/apply.py:3]: Could you use numpy's log2 instead?

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py]: `@pytest.mark.parametrize("operator", GATESET)`does ensure that the input is an element of `GATESET`. So it be rather:
```python3
@pytest.mark.parametrize("operator", GATESET)
def test_operator_product(operator: Primitive) -> None:

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py]: Same.

File: pyqtorch/apply.py
Comment (gvelikova) [inline pyqtorch/apply.py:83]: Would there be a  case where both operators are of the same size in the else?

File: pyqtorch/apply.py
Comment (gvelikova) [inline pyqtorch/apply.py:88]: Same as above.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:3]: Changed.

File: pyqtorch/apply.py
Comment (EthanObadia) [reply pyqtorch/apply.py:83]: This code part is intended to ensure that both tensors are the same `size` before `torch.bmm()` to prevent errors. If both tensors already have the same `size` it is not necessary to handle them and you can directly run `torch.bmm()`. 
This case is particularly possible when studying a 1 qubit system because the doors are mostly defined on a 1 qubit.

**Review by jpmoutinho — APPROVED** [review]
LGTM. I opened [an issue](https://github.com/pasqal-io/pyqtorch/issues/158) to keep track of the possible optimization of the operator multiplication. We can discuss it at some point.

File: tests/test_digital.py
Comment (Roland-djee) [inline tests/test_digital.py:329]: Fixtures are functions so the PEP8 convention for function names is lower case.

File: pyqtorch/primitive.py
Comment (EthanObadia) [reply pyqtorch/primitive.py:18]: Changed.

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py:328]: Changed and created `conftest.py`.

File: tests/test_digital.py
Comment (EthanObadia) [reply tests/test_digital.py:329]: Renamed `gates`.

**Pull request merged**
